core/src/main/java/org/mule/runtime/core/internal/streaming/CursorManager.java

```diff
@@ -38,152 +32,76 @@
-   *
-   * @param cursor the closed cursor
-   * @param handle the handle for the provider that generated it
-   */
-  public void onClose(Cursor cursor, CursorContext handle) {
-    final String eventId = handle.getOwnerContext().getId();
-    EventStreamingState state = registry.getIfPresent(eventId);
-
-    if (state != null && state.removeCursor(handle.getCursorProvider(), cursor)) {
-      registry.invalidate(eventId);
-    }
+  private void terminated(BaseEventContext context) {
+    registry.invalidate(context);
   }
 
-  private void terminated(BaseEventContext rootContext) {
-    EventStreamingState state = registry.getIfPresent(rootContext.getId());
-    if (state != null) {
-      registry.invalidate(rootContext.getId());
-    }
-  }
-
-  /**
-   * Duplicate registration will occur if cursors are opened in multiple child flows or processing branches. This means terminate
-   * will fire multiple times sequentially during completion of the parent EventContext. After the first terminate all other
-   * invocation will literally be no-ops. This is preferred to introducing contention here given multiple thread may be opening
-   * cursors concurrently.
-   */
-  private void registerEventContext(BaseEventContext eventContext) {
-    eventContext.onTerminated((response, throwable) -> terminated(eventContext));
+  private void hookEventTermination(BaseEventContext ownerContext) {
+    ownerContext.onTerminated((response, throwable) -> terminated(ownerContext));
   }
 
   private class EventStreamingState {
 
     private final AtomicBoolean disposed = new AtomicBoolean(false);
-    private final AtomicInteger cursorCount = new AtomicInteger(0);
-
-    private final LoadingCache<CursorProvider<CursorStream>, Set<Cursor>> cursors = Caffeine.newBuilder()
-        .removalListener((CursorProvider<CursorStream> key, Set<Cursor> value, RemovalCause cause) -> {
-          try {
-            closeProvider(key);
-            releaseAll(value);
-          } finally {
-            key.releaseResources();
-          }
-        })
-        .build(key -> {
-          statistics.incrementOpenProviders();
-          return newSetFromMap(new ConcurrentHashMap<>());
-        });
-
-    private synchronized void addProvider(CursorProvider adapter) {
-      cursors.get(adapter);
-    }
-
-    private void addCursor(CursorProvider provider, Cursor cursor) {
-      cursors.get(provider).add(cursor);
-      cursorCount.incrementAndGet();
-    }
-
-    private boolean removeCursor(CursorProvider provider, Cursor cursor) {
-      Set<Cursor> openCursors = cursors.get(provider);
-      if (openCursors.remove(cursor)) {
-        statistics.decrementOpenCursors();
-        if (cursorCount.decrementAndGet() <= 0 && provider.isClosed()) {
-          dispose();
-          return true;
-        }
-      }
+    private final Map<Integer, WeakReference<ManagedCursorProvider>> providers = new ConcurrentHashMap<>();
```

Comment:
Do you need this to be a `Map`? you do not query by the key. Wouldn't a `Set` be enough?